src/thunderbots/software/ai/primitive/directwheels_primitive.cpp

```diff
@@ -0,0 +1,129 @@
+#include "ai/primitive/directwheels_primitive.h"
+
+#include "ai/primitive/visitor/primitive_visitor.h"
+#include "util/logger/init.h"
+
+const std::string DirectWheelsPrimitive::PRIMITIVE_NAME = "DirectWheels Primitive";
+
+DirectWheelsPrimitive::DirectWheelsPrimitive(
+    unsigned int robot_id, int front_left_wheel_power, int back_left_wheel_power,
+    int front_right_wheel_power, int back_right_wheel_power, double dribbler_rpm)
+    : robot_id(robot_id),
+      front_left_wheel_power(front_left_wheel_power),
+      back_left_wheel_power(back_left_wheel_power),
+      front_right_wheel_power(front_right_wheel_power),
+      back_right_wheel_power(back_right_wheel_power),
+      dribbler_rpm(dribbler_rpm)
+{
+    // Clamp wheel power if out of range, log a warning
+    if (abs(front_left_wheel_power) > 255)
+    {
+        this->front_left_wheel_power = std::clamp(front_left_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: front left wheel power out of bounds";
+    }
+    if (abs(back_left_wheel_power) > 255)
+    {
+        this->back_left_wheel_power = std::clamp(back_left_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: back left wheel power out of bounds";
+    }
+    if (abs(front_right_wheel_power) > 255)
+    {
+        this->front_right_wheel_power = std::clamp(front_right_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: front right wheel power out of bounds";
+    }
+    if (abs(back_right_wheel_power) > 255)
+    {
+        this->back_right_wheel_power = std::clamp(back_right_wheel_power, -255, 255);
+        LOG(WARNING) << "direct wheels: back right wheel power out of bounds";
+    }
+}
+
+DirectWheelsPrimitive::DirectWheelsPrimitive(
+    const thunderbots_msgs::Primitive &primitive_msg)
+{
+    validatePrimitiveMessage(primitive_msg, getPrimitiveName());
+
+    robot_id                = primitive_msg.robot_id;
+    front_left_wheel_power  = static_cast<double>(primitive_msg.parameters.at(0));
+    back_left_wheel_power   = static_cast<double>(primitive_msg.parameters.at(1));
+    front_right_wheel_power = static_cast<double>(primitive_msg.parameters.at(2));
+    back_right_wheel_power  = static_cast<double>(primitive_msg.parameters.at(3));
+    dribbler_rpm            = static_cast<double>(primitive_msg.parameters.at(4));
```

Comment:
Instead of copying most the logic from the first constructor above into this one, could use use delegating constructors (basically calling the first constructor from the second)? https://stackoverflow.com/questions/308276/can-i-call-a-constructor-from-another-constructor-do-constructor-chaining-in-c

Reply:
This seems pretty straightforward, but there's a validatePrimitiveMessage() function in this constructor. How should I delegate the constructor such that it still keeps this function?

Reply:
Sorry, that was a poor example on my part. If you scroll down to the `Delegating Constructors` section of this page: https://docs.microsoft.com/en-us/cpp/cpp/uniform-initialization-and-delegating-constructors?view=vs-2017, there is an example where each constructor does some work before handing off to the next constructor.